ls1intum/Artemis
# [Exam] Check that submissions are made in time and the user is allowed to submit

### Checklist

### Motivation and Context
Basically we have to check the student submits in time during the exam. 

### Description
- [x] I've added `checkSubmissionAllowance()` to `SubmissionService` (see description below)
- [x] I've added `checkSubmissionAllowance()` to `ExamSubmissionService` (see description below)
- [x] I've tested the new checks thoroughly
- [x] I've added the new checks to the `TextSubmissionResource`
- [x] I've added the new checks to the `FileUploadSubmissionResource`
- [x] I've added the new checks to the `QuizSubmissionResource`
- [x] I've added the new checks to the `ProgrammingSubmissionResource` (only `@PostMapping(Constants.PROGRAMMING_SUBMISSION_RESOURCE_PATH + "{participationId}/trigger-build")`)
- [x] I've added the new checks to the `ModelingSubmissionResource`

**`SubmissionService::checkSubmissionAllowance()`**
- Make sure that the client did not change groups. We did this before in all resources, having it in one place makes testing easier
- Another check that is only applied if the user wants to update a submission: Check that the user of submission.participation is the same as the user who executes this call (with respect to teams). This prevents injecting submissions to other users.

**`ExamSubmissionService::checkSubmissionAllowance()`**
- Checks if the submission is made for an exam. If so these checks will be applied:
- Check that the current user is allowed to submit to this exercise (does the studentExam of the user have the exercise the user wants to submit to)
- Check that the submission is made in time (between `exam.startDate` and `exam.startDate + studentExam.workingTime`)

**Follow-ups:**
- Send more descriptive error messages to the client instead of just using `forbidden()`
- Add a grace period for submissions. If we do so we have to adjust the dueDate checks in the submission services (e.g. in TextSubmissionService::handleTextSubmission())

### Steps for Testing
**(1) Test exam participation**
- Participate in an exam
- Test all already supported exercise types
- Update your previously made submission at least once
- Check that you can only participate in time (between startDate of the exam and startDate+workingTime)
- Check that there is only one submission (at least locally)

**(2) Test *normal* participation**
This is not quite as important as the testing for the exam mode (1) as the checks for the exam won't be applied. As I've done smaller refactorings in all submission resources it makes sense to test the existing participation functionality.
